pkg/storage/cacher.go

```diff
@@ -59,11 +59,56 @@ type CacherConfig struct {
 	// KeyFunc is used to get a key in the underyling storage for a given object.
 	KeyFunc func(runtime.Object) (string, error)
 
+	// IndexFunc is used for optimizing amount of watchers that needs to process
+	// an incoming event.
+	IndexFunc IndexingFunc
+
 	// NewList is a function that creates new empty object storing a list of
 	// objects of type Type.
 	NewListFunc func() runtime.Object
 }
 
+type indexedWatchers struct {
+	allWatchers   map[int]*cacheWatcher
+	valueWatchers map[string]map[int]*cacheWatcher
```

Comment:
Especially considering that I think this eventually needs to be a map of index name to index value to map[int]*cacheWatcher...